File: 0210-course-schedule-ii/0210-course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
    let graph = new Map()
    let result = []
    let indegree = new Array(numCourses).fill(0)
    let queue = []
    var again = function (graph, first, second) {
        if (!graph.has(first)) {
            graph.set(first, [])
        }
        graph.get(first).push(second)
        indegree[second]++
    }

    for (let i = 0; i < prerequisites.length; i++) {
        again(graph, prerequisites[i][0], prerequisites[i][1])
    }


    for (let i = 0; i < numCourses; i++) {
        if (indegree[i] == 0) {
            queue.push(i)
        }
    }

    while (queue.length > 0) {
        let first = queue.shift()
        result.push(first)
        for (let neighbor of graph.get(first) || []) {
            indegree[neighbor]--
            if (indegree[neighbor] == 0) {
                queue.push(neighbor)
            }
        }
    }
    return result.length == numCourses ? result.reverse() : []
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = {}
        result = []
        queue = []
        indegree = [0] * numCourses

        def again(garph, first, second):
            if not first in graph:
                graph[first] = []

            graph[first].append(second)
            indegree[second] += 1

        for first, second in prerequisites:
            again(graph, first, second)

        # checking point from where we start
        for point in range(numCourses):
            if indegree[point] == 0:
                queue.append(point)

        while len(queue) > 0:
            first = queue.pop(0)
            result.append(first)
            for neighbor in graph.get(first, []):
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)

        return result[::-1] if len(result) == numCourses else []


        # Time Complexity : 0(V + E)
        # Space Complexity : 0(V + E)
